pkg/sources/adapter/azureservicebussource/adapter.go

```diff
@@ -111,33 +122,104 @@ func NewAdapter(ctx context.Context, envAcc pkgadapter.EnvConfigAccessor, ceClie
 	resID := &v1alpha1.AzureResourceID{}
 
 	err := json.Unmarshal([]byte(strconv.Quote(resIDStr)), resID)
 	if err != nil {
 		return nil, fmt.Errorf("deserializing resource ID string: %w", err)
 	}
 
+	// Must match one of the following patterns:
+	//  - /.../providers/Microsoft.ServiceBus/namespaces/{namespaceName}/queues/{queueName}
+	//  - /.../providers/Microsoft.ServiceBus/namespaces/{namespaceName}/topics/{topicName}/subscriptions/{subsName}
+	if resID.ResourceProvider != resourceProviderServiceBus ||
+		resID.Namespace == "" ||
+		resID.ResourceType != resourceTypeQueues && resID.ResourceType != resourceTypeTopics ||
+		resID.ResourceType == resourceTypeQueues && resID.SubResourceType != "" ||
+		resID.ResourceType == resourceTypeTopics && resID.SubResourceType != resourceTypeSubscriptions {
+
+		return nil, errors.New("resource ID does not refer to a Service Bus entity")
+	}
+
 	return resID, nil
 }
 
+// entityPath returns the entity path of the given Service Bus entity.
+func entityPath(entityID *v1alpha1.AzureResourceID) string {
+	switch entityID.ResourceType {
+	case resourceTypeQueues:
+		queueName := entityID.ResourceName
+		return queueName
+	case resourceTypeTopics:
+		topicName := entityID.ResourceName
+		subsName := entityID.SubResourceName
+		return topicName + "/Subscriptions/" + subsName
+	default:
+		return ""
+	}
+}
+
+// namespaceFromEnvironment mimics the behaviour of eventhub.NewHubFromEnvironment
+// by returning a servicebus.NamespaceOption that is suitable for the
+// authentication method selected via environment variables.
+func namespaceFromEnvironment(entityID *v1alpha1.AzureResourceID) servicebus.NamespaceOption {
+	return func(ns *servicebus.Namespace) error {
+		// SAS authentication (token, connection string)
+		connStr := connectionStringFromEnvironment(entityID.Namespace, entityPath(entityID))
+		sasErr := servicebus.NamespaceWithConnectionString(connStr)(ns)
+		if sasErr == nil {
+			return nil
+		}
+
+		// AAD authentication (service principal)
+		aadErr := servicebus.NamespaceWithEnvironmentBinding(entityID.Namespace)(ns)
+		if aadErr == nil {
+			return nil
+		}
+
+		return fmt.Errorf("neither Azure Active Directory nor SAS token provider could be built - "+
+			"AAD error: %v, SAS error: %v", aadErr, sasErr)
+	}
+}
+
+// connectionStringFromEnvironment returns a Service Bus connection string
+// based on values read from the environment.
+func connectionStringFromEnvironment(namespace, entityPath string) string {
+	connStr := os.Getenv(envConnStr)
+
+	// if a key is set explicitly, it takes precedence and is used to
+	// compose a new connection string
+	if keyName, keyValue := os.Getenv(envKeyName), os.Getenv(envKeyValue); keyName != "" || keyValue != "" {
+		azureEnv := &azure.PublicCloud
+		connStr = fmt.Sprintf("Endpoint=sb://%s.%s;SharedAccessKeyName=%s;SharedAccessKey=%s;EntityPath=%s",
+			namespace, azureEnv.ServiceBusEndpointSuffix, keyName, keyValue, entityPath)
+	}
+
+	return connStr
+}
```

Comment:
We can and I initially wanted to do it that way, but later had a feeling this might introduce some confusion, because the SDK itself accesses the env vars directly using `os.Getenv()`.

So for AAD auth, when we call `servicebus.NamespaceWithEnvironmentBinding()`, the function reads env vars one way, but for SAS auth, the function reads them a different way.

But I think you're right, the readability of our code shouldn't from using one SDK or another. I'll add the env vars to envconfig with a comment that explains exactly how they are being used.